src/helpers/assets.ts

```diff
@@ -236,20 +244,153 @@ export const buildUniqueTokenList = (uniqueTokens: any, selectedShowcaseTokens:
+  isShowcaseDataMigrated = false,
+  isHiddenDataMigrated = false,
+  isReadOnlyWallet = false
+) => {
+  const result: CellTypes[] = [
+    {
+      type: CellType.NFTS_HEADER,
+      uid: `nft-headers`,
+    },
+    { type: CellType.NFTS_HEADER_SPACE_AFTER, uid: 'nfts-header-space-after' },
+  ];
+
+  if (!collections?.size) {
+    if (!isFetchingNfts) {
+      result.push({ type: CellType.NFTS_EMPTY, uid: `nft-empty` });
+    } else {
+      result.push({ type: CellType.NFTS_LOADING, uid: `nft-loading` });
+    }
+  } else {
+    if (isShowcaseDataMigrated && showcaseTokens.length > 0 && listType !== 'select-nft') {
+      result.push({
+        name: i18n.t(i18n.l.account.tab_showcase),
+        total: showcaseTokens.length,
+        type: CellType.FAMILY_HEADER,
+        uid: 'showcase',
+      });
+      for (let index = 0; index < showcaseTokens.length; index++) {
+        const uniqueId = showcaseTokens[index];
+
+        const { network, contractAddress } = parseUniqueId(uniqueId);
+
+        result.push({
+          type: CellType.NFT,
+          uid: `showcase-${uniqueId}`,
+          uniqueId,
+          index,
+          collectionId: `${network}_${contractAddress}`.toLowerCase(),
+        });
+      }
+      result.push({ type: CellType.NFT_SPACE_AFTER, uid: `showcase-space-after` });
+    }
+
+    // i18n all names
+    if (sellingTokens.length > 0) {
+      result.push({
+        name: i18n.t(i18n.l.nfts.selling),
+        total: sellingTokens.length,
+        type: CellType.FAMILY_HEADER,
+        uid: 'selling',
+      });
+      for (let index = 0; index < sellingTokens.length; index++) {
+        const uniqueId = sellingTokens[index].uniqueId;
+        result.push({
+          index,
+          type: CellType.NFT,
+          uid: `selling-${uniqueId}`,
+          uniqueId,
+        });
+      }
+      result.push({ type: CellType.NFT_SPACE_AFTER, uid: `showcase-space-after` });
+    }
+
+    // Create a Map for O(1) lookups instead of filtering arrays
+    const specialTokensByCollection = new Map<string, number>();
+
+    if (isShowcaseDataMigrated && isHiddenDataMigrated) {
+      [...showcaseTokens, ...hiddenTokens].forEach(token => {
+        const { network, contractAddress } = parseUniqueId(token);
+        const collectionId = `${network}_${contractAddress}`.toLowerCase();
+        specialTokensByCollection.set(collectionId, (specialTokensByCollection.get(collectionId) || 0) + 1);
+      });
+    }
+
+    for (const { id, imageUrl, name, totalCount } of collections.values()) {
+      let adjustedTotalCount = totalCount;
+
+      if (isShowcaseDataMigrated && isHiddenDataMigrated) {
+        const specialCount = specialTokensByCollection.get(id.toLowerCase()) || 0;
+        adjustedTotalCount = String(Math.max(0, Number(totalCount) - specialCount));
+      }
+
+      if (Number(adjustedTotalCount) === 0) {
+        continue;
+      }
+
+      result.push({
+        image: imageUrl ?? undefined,
```

Comment:
?? undefined is reduntant (unless imageUrl could be  null and needs to be converted, not sure)